File: lib/src/window_manager.dart
import 'package:nativeapi/src/ffi/bindings.dart';
import 'package:nativeapi/src/ffi/bindings_generated.dart';
import 'dart:ffi';
import 'package:ffi/ffi.dart' as ffi;

import 'window.dart';

/// A class for managing windows.
///
/// This class provides methods for getting all windows and the current window.
/// It is a singleton class.
class WindowManager {
  WindowManager._();

  /// The singleton instance of the WindowManager.
  static final WindowManager instance = WindowManager._();

  /// The native API bindings.
  NativeApiBindings get _bindings => nativeApiBindings;

  /// Get all windows.
  ///
  /// Returns a list of all windows.
  List<Window> getAll() {
    final windowIdList = _bindings.window_manager_get_all();
    final List<int> idList = List.generate(
      windowIdList.count,
      (index) => windowIdList.ids[index],
    );
    ffi.malloc.free(windowIdList.ids);
    return idList.map((id) => Window(id: id)).toList();
  }

  /// Get the current window.
  ///
  /// Returns the current window.
  Window getCurrent() {
    final nativeWindowId = _bindings.window_manager_get_current();
    return Window(id: nativeWindowId);
  }
}

File: src/window_manager.cpp
#include <string.h>
#include <iostream>

#include "libnativeapi/include/nativeapi.h"
#include "window_manager.h"

using namespace nativeapi;

WindowManager g_window_manager = WindowManager();

FFI_PLUGIN_EXPORT
struct NativeWindowIDList window_manager_get_all() {
  auto windows = g_window_manager.GetAll();
  NativeWindowIDList window_id_list;
  window_id_list.count = windows.size();
  window_id_list.ids = (long*)malloc(sizeof(long) * window_id_list.count);
  for (size_t i = 0; i < window_id_list.count; i++) {
    window_id_list.ids[i] = windows[i]->id;
  }
  return window_id_list;
}

FFI_PLUGIN_EXPORT
long window_manager_get_current() {
  auto window = g_window_manager.GetCurrent();
  return window->id;
}

FFI_PLUGIN_EXPORT
void window_manager_start_listening() {
  // TODO: Implement
}

FFI_PLUGIN_EXPORT
void window_manager_stop_listening() {
  // TODO: Implement
}
